### PR TITLE
Fix `PolicySet::link` to avoid mutating policy set on link fail

### PR DESCRIPTION
**Pull request by john-h-kastner-aws**
If linking failed because the template id corresponded to a static policy, `PolicySet::link` would correctly return an error, but the error was returned after calling `ast::PolicySet::link`. When given a static policy id, this function updated the links HashMap to add an entry for the target policy id which was not removed when returning an error.

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [ ] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).
- [ ] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).
- [ ] A bug fix or other functionality change requiring a patch to `cedar-policy`.
- [ ] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)
- [ ] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [ ] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).
- [ ] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [ ] Does not require updates because my change does not impact the Cedar Dafny model or DRT infrastructure.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)
- [ ] Requires updates, but I do not plan to make them in the near future. (Make sure that your changes are hidden behind a feature flag to mark them as experimental.)
- [ ] I'm not sure how my change impacts `cedar-spec`. (Post your PR anyways, and we'll discuss in the comments.)

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cedar-policy/CHANGELOG.md [hunk 1]
```diff
@@ -64,6 +64,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   matching all other operators. This affects what error is reported when there is
   an evaluation error in both operands, but does not otherwise change the result
   of evaluation.
+- `PolicySet::link` to not mutate internal state when failing to link a static
```
Comment (andrewmwells-amazon) [inline cedar-policy/CHANGELOG.md:67]:
```suggestion
- Updated `PolicySet::link` to not mutate internal state when failing to link a static
```

File: cedar-policy/CHANGELOG.md [hunk 2]
```diff
@@ -64,6 +64,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  policy. With this fix it is be possible to create a link with a policy id
```
Comment (andrewmwells-amazon) [inline cedar-policy/CHANGELOG.md:68]:
```suggestion
  policy. With this fix it is possible to create a link with a policy id
```

**Review by andrewmwells-amazon — APPROVED** [review]

**Comment by andrewmwells-amazon** [comment]
Just out of curiosity, how was this bug discovered?

**Comment by john-h-kastner-aws** [comment]
Basically just looking at the code

**Review by shaobo-he-aws — APPROVED** [review]

**Pull request merged**